applications/nrf_desktop/src/modules/ble_discovery.c

```diff
@@ -0,0 +1,288 @@
+{
+	__ASSERT_NO_MSG(discovering_peer_conn != NULL);
+	int err = bt_unpair(BT_ID_DEFAULT,
+			    bt_conn_get_dst(discovering_peer_conn));
+
+	if (err) {
+		LOG_ERR("Error while unpairing peer (err:%d)", err);
+		module_set_state(MODULE_STATE_ERROR);
+	}
+	bt_conn_unref(discovering_peer_conn);
+	discovering_peer_conn = NULL;
+}
+
+static void hids_discovery_completed(struct bt_gatt_dm *dm, void *context)
+{
+	__ASSERT_NO_MSG(bt_gatt_dm_conn_get(dm) == discovering_peer_conn);
+	LOG_INF("HIDS discovery procedure succeeded");
+
+	bt_gatt_dm_data_print(dm);
+
+	struct ble_discovery_complete_event *event =
+		new_ble_discovery_complete_event();
+
+	event->dm = dm;
+	event->pid = peer_pid;
+
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(bt_peripherals); i++) {
+		if (bt_peripherals[i].pid == peer_pid) {
+			event->peer_type = bt_peripherals[i].peer_type;
+			break;
+		}
+	}
+
+	if (i == ARRAY_SIZE(bt_peripherals)) {
+		LOG_ERR("Unrecognized peer");
+		k_free(event);
+		int err = bt_gatt_dm_data_release(dm);
+
+		if (err) {
+			LOG_ERR("Discovery data release failed (err:%d)", err);
+			module_set_state(MODULE_STATE_ERROR);
+		}
+
+		peer_unpair();
+		return;
+	}
+	static_assert(PEER_TYPE_POS_COUNT <= 8, "");
+
+	EVENT_SUBMIT(event);
+
+	/* This module is the last one to process this event and release
+	 * the discovery data.
+	 */
+}
+
+static void hids_discovery_service_not_found(struct bt_conn *conn,
+					     void *context)
+{
+	LOG_ERR("Cannot find HIDS during the discovery");
+	__ASSERT_NO_MSG(discovering_peer_conn == conn);
+	peer_unpair();
+}
+
+static void hids_discovery_error_found(struct bt_conn *conn, int err,
+				       void *context)
+{
+	LOG_ERR("HIDS discovery failed (err:%d)", err);
+	__ASSERT_NO_MSG(discovering_peer_conn == conn);
+	peer_unpair();
+}
+
+static void start_hids_discovery(struct bt_conn *conn)
+{
+	static const struct bt_gatt_dm_cb discovery_cb = {
+		.completed = hids_discovery_completed,
+		.service_not_found = hids_discovery_service_not_found,
+		.error_found = hids_discovery_error_found,
+	};
+
+	int err = bt_gatt_dm_start(conn, BT_UUID_HIDS, &discovery_cb, NULL);
+
+	if (err == -EALREADY) {
+		LOG_ERR("Discovery already in progress");
+		module_set_state(MODULE_STATE_ERROR);
+	} else if (err) {
+		LOG_ERR("Cannot start the discovery (err:%d)", err);
+		peer_unpair();
+	}
+}
+
+static bool verify_peer(void)
+{
+	if (peer_vid != vendor_vid) {
+		LOG_ERR("Invalid peripheral VID");
+		return false;
+	}
+
+	return true;
+}
+
+static void hids_discovery_start_fn(struct k_work *w)
+{
+	if (!verify_peer()) {
+		peer_unpair();
+		return;
+	}
+	start_hids_discovery(discovering_peer_conn);
+}
+
+u8_t read_attr(struct bt_conn *conn, u8_t err,
+	       struct bt_gatt_read_params *params,
+	       const void *data, u16_t length)
+{
+	if (err) {
+		LOG_ERR("Problem with reading GATT (err:%u)", err);
+		peer_unpair();
+		return BT_GATT_ITER_STOP;
+	}
+
+	__ASSERT_NO_MSG(length >= 5);
+	u8_t *data_ptr = (u8_t *) data;
```

Comment:
explicit cast from `void *` is not needed
i.e. 
`u8_t *data_ptr = data;`

Reply:
ok